exporter/datadogexporter/translate_traces.go

```diff
@@ -0,0 +1,492 @@
+	trace.StatusCodeUnknown:            {message: "UNKNOWN", status: http.StatusInternalServerError},
+	trace.StatusCodeInvalidArgument:    {message: "INVALID_ARGUMENT", status: http.StatusBadRequest},
+	trace.StatusCodeDeadlineExceeded:   {message: "DEADLINE_EXCEEDED", status: http.StatusGatewayTimeout},
+	trace.StatusCodeNotFound:           {message: "NOT_FOUND", status: http.StatusNotFound},
+	trace.StatusCodeAlreadyExists:      {message: "ALREADY_EXISTS", status: http.StatusConflict},
+	trace.StatusCodePermissionDenied:   {message: "PERMISSION_DENIED", status: http.StatusForbidden},
+	trace.StatusCodeResourceExhausted:  {message: "RESOURCE_EXHAUSTED", status: http.StatusTooManyRequests},
+	trace.StatusCodeFailedPrecondition: {message: "FAILED_PRECONDITION", status: http.StatusBadRequest},
+	trace.StatusCodeAborted:            {message: "ABORTED", status: http.StatusConflict},
+	trace.StatusCodeOutOfRange:         {message: "OUT_OF_RANGE", status: http.StatusBadRequest},
+	trace.StatusCodeUnimplemented:      {message: "UNIMPLEMENTED", status: http.StatusNotImplemented},
+	trace.StatusCodeInternal:           {message: "INTERNAL", status: http.StatusInternalServerError},
+	trace.StatusCodeUnavailable:        {message: "UNAVAILABLE", status: http.StatusServiceUnavailable},
+	trace.StatusCodeDataLoss:           {message: "DATA_LOSS", status: http.StatusNotImplemented},
+	trace.StatusCodeUnauthenticated:    {message: "UNAUTHENTICATED", status: http.StatusUnauthorized},
+}
+
+// converts Traces into an array of datadog trace payloads grouped by env
+func ConvertToDatadogTd(td pdata.Traces, cfg *Config, globalTags []string) ([]*pb.TracePayload, error) {
+	// get hostname tag
+	// this is getting abstracted out to config
+	hostname := *GetHost(cfg)
+
+	// TODO:
+	// do we apply other global tags, like version+service, to every span or only root spans of a service
+	// should globalTags['service'] take precedence over a trace's resource.service.name? I don't believe so, need to confirm
+
+	resourceSpans := td.ResourceSpans()
+
+	traces := []*pb.TracePayload{}
+
+	for i := 0; i < resourceSpans.Len(); i++ {
+		rs := resourceSpans.At(i)
+		if rs.IsNil() {
+			continue
+		}
+
+		// TODO: Also pass in globalTags here when we know what to do with them
+		payload, err := resourceSpansToDatadogSpans(rs, hostname, cfg, globalTags)
+		if err != nil {
+			return traces, err
+		}
+
+		traces = append(traces, &payload)
+
+	}
+
+	return traces, nil
+}
+
+func AggregateTracePayloadsByEnv(tracePayloads []*pb.TracePayload) []*pb.TracePayload {
+	lookup := make(map[string]*pb.TracePayload)
+	for _, tracePayload := range tracePayloads {
+		key := fmt.Sprintf("%s|%s", tracePayload.HostName, tracePayload.Env)
+		var existingPayload *pb.TracePayload
+		if val, ok := lookup[key]; ok {
+			existingPayload = val
+		} else {
+			existingPayload = &pb.TracePayload{
+				HostName: tracePayload.HostName,
+				Env:      tracePayload.Env,
+				Traces:   make([]*pb.APITrace, 0),
```

Comment:
Would it be worth setting the initial capacity to the trace payload's traces length?

Reply:
yup definitely, updated